Allow host to to be notified of filament run out and control pause.

marcio-ao opened the pull request:
It is now possible to specify an ACTION_ON_FILAMENT_RUNOUT for the cases in which the print host (rather than Marlin) should initiate the pause.

- When ACTION_ON_FILAMENT_RUNOUT is **not** defined:
  - A filament run out will always cause FILAMENT_RUNOUT_SCRIPT to run (current Marlin behavior).
- When ACTION_ON_FILAMENT_RUNOUT is defined:
  - A filament run out while printing **from the host** will cause the host to be notified, but print resumes until the host stops sending commands.
    - Run out condition will be reset immediately after notification.
    - FILAMENT_RUNOUT_SCRIPT will **not** execute.
  - A filament run out while printing from **SD card/USB flash drive** will cause FILAMENT_RUNOUT_SCRIPT to run.

Also, enhanced ExtUI callback to specify extruder for which the run out occurred so this can be shown on the display.

InsanityAutomation commented:
Should this use the action on pause and action on resume already in config adv that's used in m600?

marcio-ao commented:
@InsanityAutomation: The motivation behind this change was for cases in which the host is in a better position to pause the printer than Marlin, so M600 would never get called.

There are all sorts of issues that crop up when Marlin attempts to pause the print itself while the host is still sending commands. The simplest solution I found, which this PR implements, is simply to let the host know that the filament has run out and the host can invoke its own pause routine (just as if the user had clicked the pause button on the host's interface). If the host really wants Marlin to initiate a filament change, it can then turn around and send a "M600" to the printer.

InsanityAutomation commented:
I understand and agree with all of that, however it still leaves the scenario where there are two separate defines that both explicitly have the function of sending a pause back to the upstream host, and in most cases will be set to the exact same string of text. Perhaps since those variables have been around since 1.1.7 they should now be moved over to the generic filament run-out section instead of the bottom of configuration_adv and simply use that same variable for either scenario, either the run out script calling m600 or this. Another item to consider here though may be the stepper timeouts. If my memory is correct you have a brake on your Z axis and homing X and Y is no big deal, so not as big of an issue for you but may be for others.

marcio-ao commented:
> I understand and agree with all of that, however it still leaves the scenario where there are two separate defines that both explicitly have the function of sending a pause back to the upstream host, and in most cases will be set to the exact same string of text.

I don't think it should be a pause string. It should be a "filament_runout" string. The host should make the decision on how to react to a filament runout. You are assuming that the correct behavior is a pause, but it could be something else (like a toolhead switch).

InsanityAutomation commented:
True, like the dynamic migration pull request that is still sitting half finished.On the other hand the only time the pause command is actually used is for filament run out and that's the expectation from other host devices such as octoprint and repetiere. I supposed that point it becomes a debate between adding arguments to the existing supported command so that it works across all or a new command and pushing plugin development to the other platforms. given your relationship with octoprint though I doubt you'll get much pushback getting them to add a new one if you want it LOL. Anyway you slice it though, The action commands probably should all be together in one location or the other not separated. If it does go with a new command, my next pull request would be to move the other two options right next to this one.

marcio-ao commented:
> On the other hand the only time the pause command is actually used is for filament run out and that's the expectation from other host devices such as octoprint and repetiere.

@InsanityAutomation: Ugh. That seems like a poor decision to assume that a filament run-out is always the same as a pause. What if the host wants to implement a different behavior on "pause" vs "filament_runout"? For example, our UI had a "Pause" and a "Filament Change" button, it would be nice if there was a way to tell the host which action was requested. Anyhow, given these different possibilities, I would say it's an argument in favor of having two different #defines for the actions. That way the user can choose to make them both be "pause", or make them different, as the need may be.

> Anyway you slice it though, The action commands probably should all be together in one location or the other not separated. If it does go with a new command, my next pull request would be to move the other two options right next to this one.

I agree that having all the actions in one section of the config file would make a lot of sense. It would be nice to be able to look in one place to see all the various actions that are defined.

InsanityAutomation commented:
If done as arguments, eg pause:filament_change 0 it would be universal was my thought. I can look up the syntax in a bit but repetiere will show unsupported commands afterwards as messages. In this case supported reason codes get processed and unimplemented falls through.

marcio-ao commented:
@InsanityAutomation: Here are the actions I found on grepping through the code:

- ACTION_ON_PAUSE:pause
- ACTION_ON_PAUSE:resume
- ACTION_ON_KILL: poweroff
- G29_ACTION_ON_FAILURE:probe_failed
- G29_ACTION_ON_RECOVER:probe_rewipe
- FILAMENT_RUNOUT_ACTION: filament_runout {extruder}

Of course, out of those, the last three were added by me...

Anyhow, I guess the question is about whether it is clearer to have them in one section, or to have them next to the feature to which they apply.

marcio-ao commented:
I think the biggest reason for having unique actions is that the host should be able to present an alert box to the user saying specifically what the error was. This is what we do with Lulzbot Cura.

InsanityAutomation commented:
And the first two were added by me! Lol Repetiere takes text after the command and displays it as a message. I honestly never tried with octoprint to see what it does. I probably should!

My thought on compatibility is that you're telling it to pause streaming regardless, then do the following action. You want streaming to pause even if the action isn't supported so by working as arguments to the base action you can get both the legacy support (it'll pause and display filament runout) with the new function actually processing that command if supported. Should minimize risk adding new sub functions underneath as well.

At minimum, pause and resume are both only used for filament run out so them going with this makes sense to me. The g29 stuff and kill command are fine as is regardless imo.

marcio-ao commented:
> If done as arguments, eg pause:filament_change 0 it would be universal was my thought.

@InsanityAutomation: I guess I am okay with using that as the message (I did not know there was a standard format for these defined somewhere), but we still would need to have a special flag that would mean "only execute FILAMENT_RUNOUT_SCRIPT when printing from the SD card". The thing is that the FILAMENT_RUNOUT_SCRIPT might be a "M600" for SD card prints, but I specifically want to have the option to suppress that script so that the host can initiate the pause.

The other thing I would maybe do is rename those #defines for consistency. So maybe this?

```
#define ACTION_ON_PAUSE "pause"
#define ACTION_ON_PAUSE "resume"
#define ACTION_ON_KILL "poweroff"
#define ACTION_ON_G29_FAILURE "probe_failed"
#define ACTION_ON_G29_RECOVER "probe_rewipe"
#define ACTION_ON_FILAMENT_RUNOUT "pause: filament_runout %d"
```

InsanityAutomation commented:
Or more #define ACTION_ON_FILAMENT_RUNOUT "ACTION_ON_PAUSE: filament_runout %d" so it just uses the above command, either in this define or at the serial echo point. 

You could also have m600 return right away if not SD printing based on the flag. Maybe use an m code to toggle the runout behavior between internal and external as a nicety since not many will have the 2 serial ports needed to do either or depending on host.

I don't think there's a documented standard anywhere, but I have spent a bit of time digging through what exists through a couple of octoprint bugs that left it not processing the resume correctly...

marcio-ao committed:
Allow host to control action on filament runout.

- It is now possible to specify a ACTION_ON_FILAMENT_RUNOUT for the cases in which the print host needs to initiate the pause.
   - When ACTION_ON_FILAMENT_RUNOUT is defined, a filament runout while printing from the host will cause the host to be notified.
   - When ACTION_ON_FILAMENT_RUNOUT is defined, a filament runout while printing from SD card will cause FILAMENT_RUNOUT_SCRIPT to run.
- Enhanced ExtUI callback to specify extruder for which the runout occurred.

marcio-ao commented:
@InsanityAutomation: I renamed the variable from FILAMENT_RUNOUT_ACTION to ACTION_ON_FILAMENT_RUNOUT and set the default value to "pause: filament_runout"

If the user does not define ACTION_ON_FILAMENT_RUNOUT, then the run-out script will execute as normal, which will call "M600" and send "//action:pause" to the host.

If the user defines ACTION_ON_FILAMENT_RUNOUT, then it will not execute the script (unless when printing from the SD card) and will output "//action:pause filament_runout N" instead. If I understand you correctly, all hosts will respond by pausing the print, which seems like an acceptable fallback behavior.

InsanityAutomation commented:
As an additional fyi, I dug this up from awhile ago - https://github.com/MarlinFirmware/Marlin/pull/11863 where some discussion was had on this topic. They used command:argument to do the same thing (RequestPause:FilamentRunout) telling repetier to do the filament runout stuff on the host side.

The issue is while repetier has the second set of commands with distinction, octoprint does not so what weve been discussing and you just changed it to is the single most universal option :)

thinkyhead committed:
Update runout.h

thinkyhead committed:
Update newest configs

thinkyhead committed:
Tweak new option descriptions

thinkyhead commented:
@repetier — This implements the "`//action:pause: filament_runout`" for filament runout on host-based printing. It will also send this string during SD-based printing — I assume this is in case it's initiated from the host side.

This all looks good to me, so it will be merged shortly. If any other refinements are needed they can follow in a new PR.

InsanityAutomation commented:
I would expect that the filament runout action being set to M25 with advanced pause disabled and the messages as described in the other ongoing thread would have achieved this. The other item is with reset being called immediately, depending on runout sensor configuration its feasible that the host would be spammed with the message as the state would be constantly reset. Should M412 be extended to reset runout status if non-blocking, and required to clear state?

repetier commented:
@thinkyhead Did I missed something or are you just ignoring the results of last discussion?

```
  // When ACTION_ON_FILAMENT_RUNOUT is declared and printing from the host (i.e, not
  // a SD print), the host will be notified and will be responsible for pausing
  // the print (the FILAMENT_RUNOUT_SCRIPT will not run in this case).
  //
  // The notification will be '//action:ACTION_ON_FILAMENT_RUNOUT <extruder_number>',
  // e.g. '//action:filament_runout 0'. The host must be configured to handle
  // the action command and execute a pause (i.e. by stopping to send commands).
  //
  //#define ACTION_ON_FILAMENT_RUNOUT "pause: filament_runout"
```

Here what I wrote in wiki upon your request:
out_of_filament T<extruder_id>: Tells host that filament for extruder extruder_id is run out or jammed. Host should issue a pause and can offer better help to user with this. For backward compatibility it should be followed by the pause action.

Also we agreed in same thread that a message after pause should be separated by a space, so the expected output would be
// action: out_of_filament T0
// action:pause filament_runout

Maybe I have missed the part as it had so many commits, but reading the commit I do not come to that output. At least the comment should use the right syntax as most users will never find the discussion.
